### ISSUE TITLE
LateInitializationError: Field '_devicePixelRatio@51203914'

### ISSUE THREAD

**Issue by appstarsbiblia** [AppLovin/AppLovin-MAX-Flutter#113]
### MAX Plugin Version

3.1.1

### Flutter Version

Apple Silicon flutter 3.10.6

### Device/Platform Info

Android

### Current Behavior

Fatal Exception: io.flutter.plugins.firebase.crashlytics.FlutterError: LateInitializationError: Field '_devicePixelRatio@51203914' has already been initialized.. Error thrown building Widget.
       at LateError._throwFieldAlreadyInitialized(dart:_internal)
       at _MaxNativeAdViewState._devicePixelRatio=(max_native_ad_view.dart:95)
       at _MaxNativeAdViewState.didChangeDependencies(max_native_ad_view.dart:130)
       at StatefulElement.performRebuild(framework.dart:5248)
       at Element.rebuild(framework.dart:4805)
       at StatefulElement.update(framework.dart:5274)
       at Element.updateChild(framework.dart:3686)
       at SingleChildRenderObjectElement.update(framework.dart:6442)
       at Element.updateChild(framework.dart:3686)
       at SingleChildRenderObjectElement.update(framework.dart:6442)
       at Element.updateChild(framework.dart:3686)
       at ComponentElement.performRebuild(framework.dart:5111)
       at Element.rebuild(framework.dart:4805)
       at StatelessElement.update(framework.dart:5162)
       at Element.updateChild(framework.dart:3686)
       at RenderObjectElement.updateChildren(framework.dart:6093)
       at MultiChildRenderObjectElement.update(framework.dart:6595)
       at Element.updateChild(framework.dart:3686)
       at SingleChildRenderObjectElement.update(framework.dart:6442)
       at Element.updateChild(framework.dart:3686)
       at SingleChildRenderObjectElement.update(framework.dart:6442)
       at Element.updateChild(framework.dart:3686)
       at SingleChildRenderObjectElement.update(framework.dart:6442)
       at Element.updateChild(framework.dart:3686)
       at SingleChildRenderObjectElement.update(framework.dart:6442)
       at Element.updateChild(framework.dart:3686)
       at SingleChildRenderObjectElement.update(framework.dart:6442)
       at Element.updateChild(framework.dart:3686)
       at SingleChildRenderObjectElement.update(framework.dart:6442)
       at Element.updateChild(framework.dart:3686)
       at ComponentElement.performRebuild(framework.dart:5111)
       at StatefulElement.performRebuild(framework.dart:5251)
       at Element.rebuild(framework.dart:4805)
       at StatefulElement.update(framework.dart:5274)
       at Element.updateChild(framework.dart:3686)
       at SingleChildRenderObjectElement.update(framework.dart:6442)
       at Element.updateChild(framework.dart:3686)
       at ComponentElement.performRebuild(framework.dart:5111)
       at Element.rebuild(framework.dart:4805)
       at ProxyElement.update(framework.dart:5417)
       at Element.updateChild(framework.dart:3686)
       at SingleChildRenderObjectElement.update(framework.dart:6442)
       at Element.updateChild(framework.dart:3686)
       at ComponentElement.performRebuild(framework.dart:5111)
       at Element.rebuild(framework.dart:4805)
       at ProxyElement.update(framework.dart:5417)
       at Element.updateChild(framework.dart:3686)
       at SingleChildRenderObjectElement.update(framework.dart:6442)
       at Element.updateChild(framework.dart:3686)
       at SingleChildRenderObjectElement.update(framework.dart:6442)
       at Element.updateChild(framework.dart:3686)
       at SingleChildRenderObjectElement.update(framework.dart:6442)
       at Element.updateChild(framework.dart:3686)
       at ComponentElement.performRebuild(framework.dart:5111)
       at Element.rebuild(framework.dart:4805)
       at ProxyElement.update(framework.dart:5417)
       at Element.updateChild(framework.dart:3686)
       at ComponentElement.performRebuild(framework.dart:5111)
       at StatefulElement.performRebuild(framework.dart:5251)
       at Element.rebuild(framework.dart:4805)
       at StatefulElement.update(framework.dart:5274)
       at Element.updateChild(framework.dart:3686)
       at ComponentElement.performRebuild(framework.dart:5111)
       at StatefulElement.performRebuild(framework.dart:5251)
       at Element.rebuild(framework.dart:4805)
       at StatefulElement.update(framework.dart:5274)
       at Element.updateChild(framework.dart:3686)
       at ComponentElement.performRebuild(framework.dart:5111)
       at Element.rebuild(framework.dart:4805)
       at ProxyElement.update(framework.dart:5417)
       at Element.updateChild(framework.dart:3686)
       at ComponentElement.performRebuild(framework.dart:5111)
       at Element.rebuild(framework.dart:4805)
       at StatelessElement.update(framework.dart:5162)
       at Element.updateChild(framework.dart:3686)
       at SingleChildRenderObjectElement.update(framework.dart:6442)
       at Element.updateChild(framework.dart:3686)
       at SingleChildRenderObjectElement.update(framework.dart:6442)
       at Element.updateChild(framework.dart:3686)
       at ComponentElement.performRebuild(framework.dart:5111)
       at StatefulElement.performRebuild(framework.dart:5251)
       at Element.rebuild(framework.dart:4805)
       at StatefulElement.update(framework.dart:5274)
       at Element.updateChild(framework.dart:3686)
       at ComponentElement.performRebuild(framework.dart:5111)
       at Element.rebuild(framework.dart:4805)
       at ProxyElement.update(framework.dart:5417)
       at Element.updateChild(framework.dart:3686)
       at ComponentElement.performRebuild(framework.dart:5111)
       at Element.rebuild(framework.dart:4805)
       at ProxyElement.update(framework.dart:5417)
       at Element.updateChild(framework.dart:3686)
       at ComponentElement.performRebuild(framework.dart:5111)
       at StatefulElement.performRebuild(framework.dart:5251)
       at Element.rebuild(framework.dart:4805)
       at StatefulElement.update(framework.dart:5274)
       at Element.updateChild(framework.dart:3686)
       at ComponentElement.performRebuild(framework.dart:5111)
       at Element.rebuild(framework.dart:4805)
       at StatelessElement.update(framework.dart:5162)
       at Element.updateChild(framework.dart:3686)
       at SliverMultiBoxAdaptorElement.updateChild(sliver.dart:857)
       at SliverMultiBoxAdaptorElement.performRebuild.processElement(sliver.dart:758)
       at Iterable.forEach(dart:core)
       at SliverMultiBoxAdaptorElement.performRebuild(sliver.dart:805)
       at SliverMultiBoxAdaptorElement.update(sliver.dart:734)
       at Element.updateChild(framework.dart:3686)
       at ComponentElement.performRebuild(framework.dart:5111)
       at Element.rebuild(framework.dart:4805)
       at ProxyElement.update(framework.dart:5417)
       at Element.updateChild(framework.dart:3686)
       at SingleChildRenderObjectElement.update(framework.dart:6442)
       at Element.updateChild(framework.dart:3686)
       at RenderObjectElement.updateChildren(framework.dart:6093)
       at MultiChildRenderObjectElement.update(framework.dart:6595)
       at Element.updateChild(framework.dart:3686)
       at SingleChildRenderObjectElement.update(framework.dart:6442)
       at Element.updateChild(framework.dart:3686)
       at SingleChildRenderObjectElement.update(framework.dart:6442)
       at Element.updateChild(framework.dart:3686)
       at SingleChildRenderObjectElement.update(framework.dart:6442)
       at Element.updateChild(framework.dart:3686)
       at SingleChildRenderObjectElement.update(framework.dart:6442)
       at Element.updateChild(framework.dart:3686)
       at ComponentElement.performRebuild(framework.dart:5111)
       at StatefulElement.performRebuild(framework.dart:5251)
       at Element.rebuild(framework.dart:4805)
       at StatefulElement.update(framework.dart:5274)
       at Element.updateChild(framework.dart:3686)
       at SingleChildRenderObjectElement.update(framework.dart:6442)
       at Element.updateChild(framework.dart:3686)
       at ComponentElement.performRebuild(framework.dart:5111)
       at Element.rebuild(framework.dart:4805)
       at ProxyElement.update(framework.dart:5417)
       at Element.updateChild(framework.dart:3686)
       at SingleChildRenderObjectElement.update(framework.dart:6442)
       at Element.updateChild(framework.dart:3686)
       at ComponentElement.performRebuild(framework.dart:5111)
       at Element.rebuild(framework.dart:4805)
       at ProxyElement.update(framework.dart:5417)
       at Element.updateChild(framework.dart:3686)
       at SingleChildRenderObjectElement.update(framework.dart:6442)
       at Element.updateChild(framework.dart:3686)
       at SingleChildRenderObjectElement.update(framework.dart:6442)
       at Element.updateChild(framework.dart:3686)
       at SingleChildRenderObjectElement.update(framework.dart:6442)
       at Element.updateChild(framework.dart:3686)
       at ComponentElement.performRebuild(framework.dart:5111)
       at Element.rebuild(framework.dart:4805)
       at ProxyElement.update(framework.dart:5417)
       at Element.updateChild(framework.dart:3686)
       at ComponentElement.performRebuild(framework.dart:5111)
       at StatefulElement.performRebuild(framework.dart:5251)
       at Element.rebuild(framework.dart:4805)
       at StatefulElement.update(framework.dart:5274)
       at Element.updateChild(framework.dart:3686)
       at ComponentElement.performRebuild(framework.dart:5111)
       at StatefulElement.performRebuild(framework.dart:5251)
       at Element.rebuild(framework.dart:4805)
       at StatefulElement.update(framework.dart:5274)
       at Element.updateChild(framework.dart:3686)
       at ComponentElement.performRebuild(framework.dart:5111)
       at Element.rebuild(framework.dart:4805)
       at StatelessElement.update(framework.dart:5162)
       at Element.updateChild(framework.dart:3686)
       at ComponentElement.performRebuild(framework.dart:5111)
       at Element.rebuild(framework.dart:4805)
       at ProxyElement.update(framework.dart:5417)
       at Element.updateChild(framework.dart:3686)
       at RenderObjectElement.updateChildren(framework.dart:6093)
       at MultiChildRenderObjectElement.update(framework.dart:6595)
       at Element.updateChild(framework.dart:3686)
       at RenderObjectElement.updateChildren(framework.dart:6093)
       at MultiChildRenderObjectElement.update(framework.dart:6595)
       at Element.updateChild(framework.dart:3686)
       at ComponentElement.performRebuild(framework.dart:5111)
       at Element.rebuild(framework.dart:4805)
       at StatelessElement.update(framework.dart:5162)
       at Element.updateChild(framework.dart:3686)
       at ComponentElement.performRebuild(framework.dart:5111)
       at Element.rebuild(framework.dart:4805)
       at StatelessElement.update(framework.dart:5162)
       at Element.updateChild(framework.dart:3686)
       at ComponentElement.performRebuild(framework.dart:5111)
       at Element.rebuild(framework.dart:4805)
       at StatelessElement.update(framework.dart:5162)
       at Element.updateChild(framework.dart:3686)
       at ComponentElement.performRebuild(framework.dart:5111)
       at Element.rebuild(framework.dart:4805)
       at ProxyElement.update(framework.dart:5417)
       at Element.updateChild(framework.dart:3686)
       at ComponentElement.performRebuild(framework.dart:5111)
       at Element.rebuild(framework.dart:4805)
       at ProxyElement.update(framework.dart:5417)
       at Element.updateChild(framework.dart:3686)
       at RenderObjectElement.updateChildren(framework.dart:6093)
       at MultiChildRenderObjectElement.update(framework.dart:6595)
       at Element.updateChild(framework.dart:3686)
       at ComponentElement.performRebuild(framework.dart:5111)
       at Element.rebuild(framework.dart:4805)
       at ProxyElement.update(framework.dart:5417)
       at Element.updateChild(framework.dart:3686)
       at ComponentElement.performRebuild(framework.dart:5111)
       at StatefulElement.performRebuild(framework.dart:5251)
       at Element.rebuild(framework.dart:4805)
       at StatefulElement.update(framework.dart:5274)
       at Element.updateChild(framework.dart:3686)
       at ComponentElement.performRebuild(framework.dart:5111)
       at StatefulElement.performRebuild(framework.dart:5251)
       at Element.rebuild(framework.dart:4805)
       at StatefulElement.update(framework.dart:5274)
       at Element.updateChild(framework.dart:3686)
       at ComponentElement.performRebuild(framework.dart:5111)
       at Element.rebuild(framework.dart:4805)
       at ProxyElement.update(framework.dart:5417)
       at Element.updateChild(framework.dart:3686)
       at ComponentElement.performRebuild(framework.dart:5111)
       at StatefulElement.performRebuild(framework.dart:5251)
       at Element.rebuild(framework.dart:4805)
       at StatefulElement.update(framework.dart:5274)
       at Element.updateChild(framework.dart:3686)
       at SingleChildRenderObjectElement.update(framework.dart:6442)
       at Element.updateChild(framework.dart:3686)
       at ComponentElement.performRebuild(framework.dart:5111)
       at Element.rebuild(framework.dart:4805)
       at ProxyElement.update(framework.dart:5417)
       at Element.updateChild(framework.dart:3686)
       at SingleChildRenderObjectElement.update(framework.dart:6442)
       at Element.updateChild(framework.dart:3686)
       at ComponentElement.performRebuild(framework.dart:5111)
       at StatefulElement.performRebuild(framework.dart:5251)
       at Element.rebuild(framework.dart:4805)
       at StatefulElement.update(framework.dart:5274)
       at Element.updateChild(framework.dart:3686)
       at ComponentElement.performRebuild(framework.dart:5111)
       at StatefulElement.performRebuild(framework.dart:5251)
       at Element.rebuild(framework.dart:4805)
       at StatefulElement.update(framework.dart:5274)
       at Element.updateChild(framework.dart:3686)
       at ComponentElement.performRebuild(framework.dart:5111)
       at Element.rebuild(framework.dart:4805)
       at ProxyElement.update(framework.dart:5417)
       at Element.updateChild(framework.dart:3686)
       at ComponentElement.performRebuild(framework.dart:5111)
       at Element.rebuild(framework.dart:4805)
       at ProxyElement.update(framework.dart:5417)
       at Element.updateChild(framework.dart:3686)
       at ComponentElement.performRebuild(framework.dart:5111)
       at Element.rebuild(framework.dart:4805)
       at ProxyElement.update(framework.dart:5417)
       at Element.updateChild(framework.dart:3686)
       at ComponentElement.performRebuild(framework.dart:5111)
       at StatefulElement.performRebuild(framework.dart:5251)
       at Element.rebuild(framework.dart:4805)
       at StatefulElement.update(framework.dart:5274)
       at Element.updateChild(framework.dart:3686)
       at ComponentElement.performRebuild(framework.dart:5111)
       at Element.rebuild(framework.dart:4805)
       at ProxyElement.update(framework.dart:5417)
       at Element.updateChild(framework.dart:3686)
       at ComponentElement.performRebuild(framework.dart:5111)
       at StatefulElement.performRebuild(framework.dart:5251)
       at Element.rebuild(framework.dart:4805)
       at StatefulElement.update(framework.dart:5274)
       at Element.updateChild(framework.dart:3686)
       at ComponentElement.performRebuild(framework.dart:5111)
       at Element.rebuild(framework.dart:4805)
       at ProxyElement.update(framework.dart:5417)
       at Element.updateChild(framework.dart:3686)
       at SingleChildRenderObjectElement.update(framework.dart:6442)
       at Element.updateChild(framework.dart:3686)
       at ComponentElement.performRebuild(framework.dart:5111)
       at Element.rebuild(framework.dart:4805)
       at StatelessElement.update(framework.dart:5162)
       at Element.updateChild(framework.dart:3686)
       at ComponentElement.performRebuild(framework.dart:5111)
       at StatefulElement.performRebuild(framework.dart:5251)
       at Element.rebuild(framework.dart:4805)
       at BuildOwner.buildScope(framework.dart:2780)
       at WidgetsBinding.drawFrame(binding.dart:903)
       at RendererBinding._handlePersistentFrameCallback(binding.dart:358)
       at SchedulerBinding._invokeFrameCallback(binding.dart:1284)
       at SchedulerBinding.handleDrawFrame(binding.dart:1214)
       at SchedulerBinding._handleDrawFrame(binding.dart:1072)

### Expected Behavior

There's an issue with the max_native_ad_view file that causes a crash. The solution I found was to fix the file to look like this:

late final double _devicePixelRatio;
bool _isDevicePixelRatioInitialized = false;

@override
void didChangeDependencies() {
  super.didChangeDependencies();
  if (!_isDevicePixelRatioInitialized && defaultTargetPlatform == TargetPlatform.android) {
    MediaQueryData queryData = MediaQuery.of(context);
    _devicePixelRatio = queryData.devicePixelRatio;
    _isDevicePixelRatioInitialized = true;
  }
}

### How to Reproduce

nativoApplovin ? Container(
                                              margin: const EdgeInsets.all(0.0),
                                              height: 300,
                                              child: applovin.MaxNativeAdView(
                                                adUnitId: AdmInterstitial.getNativoApplovinAdUnitId(),
                                                controller: _nativeAdViewController,
                                                listener: applovin.NativeAdListener(onAdLoadedCallback: (ad) {
                                                  logStatus('Native ad loaded from ${ad.networkName}');
                                                  setState(() {
                                                    _mediaViewAspectRatio = ad.nativeAd?.mediaContentAspectRatio ?? _kMediaViewAspectRatio;
                                                  });
                                                }, onAdLoadFailedCallback: (adUnitId, error) {
                                                  logStatus('Native ad failed to load with error code ${error.code} and message: ${error.message}');
                                                }, onAdClickedCallback: (ad) {
                                                  logStatus('Native ad clicked');
                                                }, onAdRevenuePaidCallback: (ad) {
                                                  logStatus('Native ad revenue paid: ${ad.revenue}');
                                                }),
                                                child: Container(
                                                  color: const Color(0xFF000000),
                                                  padding: const EdgeInsets.all(8.0),
                                                  child: Column(
                                                    mainAxisSize: MainAxisSize.min,
                                                    children: [
                                                      Row(
                                                        mainAxisAlignment: MainAxisAlignment.center,
                                                        children: [
                                                          Container(
                                                            padding: const EdgeInsets.all(4.0),
                                                            child: const applovin.MaxNativeAdIconView(
                                                              width: 48,
                                                              height: 48,
                                                            ),
                                                          ),
                                                          Flexible(
                                                            child: Column(
                                                              mainAxisAlignment: MainAxisAlignment.start,
                                                              crossAxisAlignment: CrossAxisAlignment.start,
                                                              children: [
                                                                applovin.MaxNativeAdTitleView(
                                                                  style: TextStyle(fontWeight: FontWeight.bold, fontSize: 16, color: Colors.white),
                                                                  maxLines: 1,
                                                                  overflow: TextOverflow.visible,
                                                                ),
                                                                applovin.MaxNativeAdAdvertiserView(
                                                                  style: TextStyle(fontWeight: FontWeight.normal, fontSize: 10, color: Colors.white),
                                                                  maxLines: 1,
                                                                  overflow: TextOverflow.fade,
                                                                ),
                                                                applovin.MaxNativeAdStarRatingView(
                                                                  color: Colors.white,
                                                                  size: 10,
                                                                ),
                                                              ],
                                                            ),
                                                          ),
                                                          const applovin.MaxNativeAdOptionsView(
                                                            width: 20,
                                                            height: 20,
                                                          ),
                                                        ],
                                                      ),
                                                      Row(
                                                        mainAxisAlignment: MainAxisAlignment.start,
                                                        children: [
                                                          Flexible(
                                                            child: applovin.MaxNativeAdBodyView(
                                                              style: TextStyle(fontWeight: FontWeight.normal, fontSize: 14, color: Colors.white),
                                                              maxLines: 3,
                                                              overflow: TextOverflow.ellipsis,
                                                            ),
                                                          ),
                                                        ],
                                                      ),
                                                      const SizedBox(height: 8),
                                                      Expanded(
                                                        child: AspectRatio(
                                                          aspectRatio: _mediaViewAspectRatio,
                                                          child: const applovin.MaxNativeAdMediaView(),
                                                        ),
                                                      ),
                                                      const SizedBox(
                                                        width: double.infinity,
                                                        child: applovin.MaxNativeAdCallToActionView(
                                                          style: ButtonStyle(
                                                            backgroundColor: MaterialStatePropertyAll<Color>(Color(0xFF004D40)),
                                                            textStyle: MaterialStatePropertyAll<TextStyle>(TextStyle(fontSize: 20, fontWeight: FontWeight.bold, color: Colors.white)),
                                                          ),
                                                        ),
                                                      ),
                                                    ],
                                                  ),
                                                ),
                                              ),
                                            ):Container()

**Comment by alhiwatan** [comment]
@appstarsbiblia, thanks for raising this with the solution.  

Since `_devicePixelRatio` is used only in `_addNativeAdComponent()` and `MediaQuery.devicePixelRatioOf()` is a static getter, I'd like to go with this, removing `didChangeDependencies()`.

```
  void _addNativeAdComponent(GlobalKey? key, String method) {
    ...
    if (defaultTargetPlatform == TargetPlatform.android) {
      double devicePixelRatio = MediaQuery.devicePixelRatioOf(context);
      _methodChannel!.invokeMethod(method, <String, dynamic>{
        'x': (rect.left * devicePixelRatio).round(),
        'y': (rect.top * devicePixelRatio).round(),
        'width': (rect.width * devicePixelRatio).round(),
        'height': (rect.height * devicePixelRatio).round(),
      });
    } else if (defaultTargetPlatform == TargetPlatform.iOS) {
      ...
```

Please confirm this on your side.

**Comment by appstarsbiblia** [comment]
Hello team, I will wait for the error update on the pub. Thank you for your support!

**Comment by alhiwatan** [comment]
@appstarsbiblia, we have released the fix in v3.1.2.
